In [2]:
import pandas as pd
import os
import numpy as np
from umls_api import API
from dotenv import load_dotenv
import pickle
import itertools
from collections import Counter
from tqdm.notebook import tqdm
import joblib 


load_dotenv()

True

In [3]:
# Get a list of all the cuis
# df_cuis = pd.concat([pd.read_csv("..\\data\\raw\\train_concepts.csv").set_index("ID"),pd.read_csv("..\\data\\raw\\test_concepts.csv").set_index("ID")])
df_cuis = pd.read_csv("..\\data\\raw\\cui_mapping.csv")
xss = [cuis.split(";") for cuis in df_cuis["CUI"]]
cuis = [x for xs in xss for x in xs]

In [ ]:


cuis_semantic_types = dict()

def get_cui_semantic_types(cui):
    try:
        api = API(api_key=os.getenv("UMLS_API_KEY"))  # Initialize once
        resp = api.get_cui(cui=cui)
        cui_semantic_types = resp["result"]["semanticTypes"]
        names = [st["name"] for st in cui_semantic_types]
        return {cui: set(names)}
    except Exception as e:
        print(f"Error processing CUI {cui}: {e}")
        return {cui: set()}


results = joblib.Parallel(n_jobs=-1, verbose=10)(
    joblib.delayed(get_cui_semantic_types)(cui) 
    for cui in tqdm(cuis, desc="Processing CUIs")
)
    
cuis_semantic_types = dict()
for d in results:
    key = [*d][0]
    cuis_semantic_types[key] = d[key]

with open("..\\data\\interim\\cuis_semantic_types.pkl", "wb") as f:
    pickle.dump(cuis_semantic_types, f)


Processing CUIs:   0%|          | 0/1935 [00:00<?, ?it/s]

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 20 concurrent workers.
[Parallel(n_jobs=-1)]: Done   1 tasks      | elapsed:    4.6s
[Parallel(n_jobs=-1)]: Done  10 tasks      | elapsed:    6.0s
[Parallel(n_jobs=-1)]: Done  21 tasks      | elapsed:    9.7s
[Parallel(n_jobs=-1)]: Done  32 tasks      | elapsed:    9.9s
[Parallel(n_jobs=-1)]: Done  45 tasks      | elapsed:   14.8s
[Parallel(n_jobs=-1)]: Done  58 tasks      | elapsed:   15.1s
[Parallel(n_jobs=-1)]: Done  73 tasks      | elapsed:   19.2s
[Parallel(n_jobs=-1)]: Done  88 tasks      | elapsed:   25.1s
[Parallel(n_jobs=-1)]: Done 105 tasks      | elapsed:   29.9s
[Parallel(n_jobs=-1)]: Done 122 tasks      | elapsed:   36.0s
[Parallel(n_jobs=-1)]: Done 141 tasks      | elapsed:   40.8s
[Parallel(n_jobs=-1)]: Done 160 tasks      | elapsed:   44.2s
[Parallel(n_jobs=-1)]: Done 181 tasks      | elapsed:   53.3s
[Parallel(n_jobs=-1)]: Done 202 tasks      | elapsed:   58.1s
[Parallel(n_jobs=-1)]: Done 225 tasks      | elapsed:  

{'C0016522': {'Congenital Abnormality', 'Disease or Syndrome'}, 'C0457846': {'Body Part, Organ, or Organ Component'}, 'C0578537': {'Finding'}, 'C0179751': {'Medical Device'}, 'C3489393': {'Acquired Abnormality'}, 'C0730166': {'Body Part, Organ, or Organ Component'}, 'C1261077': {'Body Part, Organ, or Organ Component'}, 'C0152443': {'Disease or Syndrome'}, 'C0227301': {'Body Part, Organ, or Organ Component'}, 'C0031354': {'Body Part, Organ, or Organ Component'}, 'C4039172': {'Body Location or Region'}, 'C0227387': {'Body Location or Region'}, 'C0227385': {'Body Part, Organ, or Organ Component'}, 'C0228126': {'Body Part, Organ, or Organ Component'}, 'C0230436': {'Body Location or Region'}, 'C0037922': {'Body Space or Junction'}, 'C1260954': {'Anatomical Abnormality'}, 'C0450184': {'Body Part, Organ, or Organ Component'}, 'C0729936': {'Body Part, Organ, or Organ Component'}, 'C0032718': {'Body Part, Organ, or Organ Component'}, 'C0149543': {'Body Location or Region'}, 'C0027540': {'Pathol

In [76]:

with open("..\\data\\interim\\cuis_semantic_types.pkl", "wb") as f:
    pickle.dump(cuis_semantic_types, f)
    
cuis_semantic_types

{'C0016522': {'Congenital Abnormality', 'Disease or Syndrome'},
 'C0457846': {'Body Part, Organ, or Organ Component'},
 'C0578537': {'Finding'},
 'C0179751': {'Medical Device'},
 'C3489393': {'Acquired Abnormality'},
 'C0730166': {'Body Part, Organ, or Organ Component'},
 'C1261077': {'Body Part, Organ, or Organ Component'},
 'C0152443': {'Disease or Syndrome'},
 'C0227301': {'Body Part, Organ, or Organ Component'},
 'C0031354': {'Body Part, Organ, or Organ Component'},
 'C4039172': {'Body Location or Region'},
 'C0227387': {'Body Location or Region'},
 'C0227385': {'Body Part, Organ, or Organ Component'},
 'C0228126': {'Body Part, Organ, or Organ Component'},
 'C0230436': {'Body Location or Region'},
 'C0037922': {'Body Space or Junction'},
 'C1260954': {'Anatomical Abnormality'},
 'C0450184': {'Body Part, Organ, or Organ Component'},
 'C0729936': {'Body Part, Organ, or Organ Component'},
 'C0032718': {'Body Part, Organ, or Organ Component'},
 'C0149543': {'Body Location or Region'},


In [ ]:
cuis_semantic_types

{'C0016522': {'Congenital Abnormality', 'Disease or Syndrome'},
 'C0457846': {'Body Part, Organ, or Organ Component'},
 'C0578537': {'Finding'},
 'C0179751': {'Medical Device'},
 'C3489393': {'Acquired Abnormality'},
 'C0730166': {'Body Part, Organ, or Organ Component'},
 'C1261077': {'Body Part, Organ, or Organ Component'},
 'C0152443': {'Disease or Syndrome'},
 'C0227301': {'Body Part, Organ, or Organ Component'},
 'C0031354': {'Body Part, Organ, or Organ Component'},
 'C4039172': {'Body Location or Region'},
 'C0227387': {'Body Location or Region'},
 'C0227385': {'Body Part, Organ, or Organ Component'},
 'C0228126': {'Body Part, Organ, or Organ Component'},
 'C0230436': {'Body Location or Region'},
 'C0037922': {'Body Space or Junction'},
 'C1260954': {'Anatomical Abnormality'},
 'C0450184': {'Body Part, Organ, or Organ Component'},
 'C0729936': {'Body Part, Organ, or Organ Component'},
 'C0032718': {'Body Part, Organ, or Organ Component'},
 'C0149543': {'Body Location or Region'},


In [59]:
for cui, st in cuis_semantic_types.items():
    if len(st) == 0:
        print(f"CUI {cui} has no semantic types.")

CUI C0241790 has no semantic types.
CUI C0021818 has no semantic types.
CUI C1134719 has no semantic types.
CUI C0206702 has no semantic types.


In [ ]:
# CUI C0241790 has no semantic types.
# CUI C0021818 has no semantic types.
# CUI C1134719 has no semantic types.
# CUI C0206702 has no semantic types.

In [69]:
cuis_semantic_types["C0241790"] = {'Congenital Abnormality'}
cuis_semantic_types["C0021818"] = {'Disease or Syndrome'}
cuis_semantic_types["C1134719"] = {'Neoplastic Process'}
cuis_semantic_types["C0206702"] = {'Neoplastic Process'}


In [70]:
semantic_types_freq = {k:v for k,v in Counter(itertools.chain.from_iterable(cuis_semantic_types.values())).items()}
semantic_types_freq

{'Congenital Abnormality': 53,
 'Disease or Syndrome': 271,
 'Body Part, Organ, or Organ Component': 809,
 'Finding': 2,
 'Medical Device': 43,
 'Acquired Abnormality': 26,
 'Body Location or Region': 157,
 'Body Space or Junction': 115,
 'Anatomical Abnormality': 59,
 'Pathologic Function': 119,
 'Substance': 2,
 'Neoplastic Process': 98,
 'Diagnostic Procedure': 27,
 'Functional Concept': 40,
 'Qualitative Concept': 21,
 'Tissue': 38,
 'Organ or Tissue Function': 8,
 'Manufactured Object': 4,
 'Body Substance': 26,
 'Spatial Concept': 4,
 'Organism Function': 6,
 'Body System': 9}

In [72]:
df = pd.read_csv("..\\data\\raw\\train_concepts.csv")
df.head()

ID                                 CUIs
0  ROCOv2_2023_train_000001                             C0040405
1  ROCOv2_2023_train_000002                             C0041618
2  ROCOv2_2023_train_000003           C0040405;C0817096;C0205271
3  ROCOv2_2023_train_000004  C0002978;C0036033;C1266909;C0225317
4  ROCOv2_2023_train_000005           C0040405;C0817096;C0497156

Disease or Syndrome (271) → è il più importante, rappresenta direttamente malattie e sindromi.<br>

Neoplastic Process (98) → include tumori e neoplasie, molto rilevante in radiologia.<br>

Pathologic Function (119) → processi patologici generici (es. infiammazione).<br>

Congenital Abnormality (53) → malformazioni congenite.<br>

Acquired Abnormality (26) → anomalie acquisite (traumi, degenerazioni).<br>

Anatomical Abnormality (59) → strutture anatomiche anormali (es. aneurisma).<br>

In [75]:
relevant_st = {'Disease or Syndrome': 0, 
               'Neoplastic Process': 0, 
               'Pathologic Function': 0, 
               'Congenital Abnormality': 0, 
               'Acquired Abnormality': 0, 
               'Anatomical Abnormality': 0
               }

for cuis in df["CUIs"]:
    cuis = cuis.split(";")
    for cui in cuis:
        if cui not in cuis_semantic_types:
            continue
        st = cuis_semantic_types[cui]
        for semantic in st:
            if semantic in relevant_st:
                relevant_st[semantic] += 1

relevant_st
    

{'Disease or Syndrome': 7676,
 'Neoplastic Process': 3427,
 'Pathologic Function': 11213,
 'Congenital Abnormality': 902,
 'Acquired Abnormality': 828,
 'Anatomical Abnormality': 2857}

Disease or Syndrome + Neoplastic Process + Anatomical Abnormality

In [32]:
important_st = ['Disease or Syndrome', 'Neoplastic Process', 'Anatomical Abnormality']

def get_dataframe(split):
    cuis_semantic_types = pickle.load(open("..\\data\\interim\\cuis_semantic_types.pkl", "rb"))

    base_dir = "..\\data\\raw\\"
    df = pd.read_csv(os.path.join(base_dir, f"{split}_concepts.csv"))
    
    print(f"Initial shape: {df.shape}")
    # split cuis in a list
    df["CUIs"] = df["CUIs"].apply(lambda x: x.split(";"))
    
    # get cuis of interest (Disease or Syndrome, Neoplastic Process, Anatomical Abnormality)
    def filter_cuis(cuis):
        filtered = [
            cui 
            for cui in cuis 
            if cui in cuis_semantic_types
            for semantic in cuis_semantic_types[cui] 
            if semantic in important_st
        ]
        return filtered if filtered else None
    
    df["CUIs"] = df["CUIs"].apply(filter_cuis)
    
    # eliminate empty cuis
    df = df.dropna(subset=["CUIs"])
    print(f"Shape after filtering CUIs: {df.shape}")
    
    # get captions
    df_captions = pd.read_csv(os.path.join(base_dir, f"{split}_captions.csv")).set_index("ID")
    df = df.set_index("ID").join(df_captions, how="inner").reset_index()
    print(f"Shape after joining with captions: {df.shape}")
    
    # get images
    df_images = pd.DataFrame()
    image_files = os.listdir(os.path.join(base_dir, split))
    df_images["Image"] = image_files
    df_images["ID"] = df_images["Image"].apply(lambda x: x.split(".")[0])
    
    df = df.set_index("ID").join(df_images.set_index("ID"), how="inner").reset_index()
    print(f"Shape after joining with images: {df.shape}")
    
    return df

df = get_dataframe("train")
df.to_pickle("..\\data\\processed\\train_processed.pkl")

Initial shape: (59958, 2)
Shape after filtering CUIs: (12133, 2)
Shape after joining with captions: (12133, 3)
Shape after joining with images: (12130, 4)
